File: web/canvas/adjustment_utils.js
/**
 * adjustment_utils.js
 *
 * 统一的图像调节工具模块，提供参数范围定义、验证和转换函数。
 * 确保前后端使用一致的参数范围和算法。
 */

// 统一的参数范围定义（与后端保持一致）
export const ADJUSTMENT_RANGES = {
  brightness: { min: -1.0, max: 1.0, default: 0.0, step: 0.01 },
  contrast: { min: -100.0, max: 100.0, default: 0.0, step: 0.1 },
  saturation: { min: -100.0, max: 100.0, default: 0.0, step: 0.1 },
  opacity: { min: 0.0, max: 100.0, default: 100.0, step: 0.1 }
};

// 亮度滑块转换因子（前端滑块范围：-100 到 100，对应亮度值：-1.0 到 1.0）
export const BRIGHTNESS_SLIDER_FACTOR = 100;

/**
 * 限制值在指定范围内
 * @param {number} value - 输入值
 * @param {number} min - 最小值
 * @param {number} max - 最大值
 * @param {number} fallback - 无效值时的默认值
 * @returns {number} 限制后的值
 */
export function clamp(value, min, max, fallback = 0) {
  const number = Number(value);
  if (Number.isNaN(number) || !Number.isFinite(number)) {
    return fallback;
  }
  return Math.min(Math.max(number, min), max);
}

/**
 * 规范化调节状态，确保所有调节参数在有效范围内
 * @param {Object} state - 原始调节状态
 * @returns {Object} 规范化后的调节状态
 */
export function normalizeAdjustmentState(state = {}) {
  return {
    brightness: clamp(
      state.brightness ?? ADJUSTMENT_RANGES.brightness.default,
      ADJUSTMENT_RANGES.brightness.min,
      ADJUSTMENT_RANGES.brightness.max,
      ADJUSTMENT_RANGES.brightness.default
    ),
    contrast: clamp(
      state.contrast ?? ADJUSTMENT_RANGES.contrast.default,
      ADJUSTMENT_RANGES.contrast.min,
      ADJUSTMENT_RANGES.contrast.max,
      ADJUSTMENT_RANGES.contrast.default
    ),
    saturation: clamp(
      state.saturation ?? ADJUSTMENT_RANGES.saturation.default,
      ADJUSTMENT_RANGES.saturation.min,
      ADJUSTMENT_RANGES.saturation.max,
      ADJUSTMENT_RANGES.saturation.default
    ),
    opacity: clamp(
      state.opacity ?? ADJUSTMENT_RANGES.opacity.default,
      ADJUSTMENT_RANGES.opacity.min,
      ADJUSTMENT_RANGES.opacity.max,
      ADJUSTMENT_RANGES.opacity.default
    )
  };
}

/**
 * 将亮度值转换为滑块值
 * @param {number} brightness - 亮度值（-1.0 到 1.0）
 * @returns {number} 滑块值（-100 到 100）
 */
export function brightnessToSlider(brightness) {
  const clamped = clamp(brightness, -1, 1, 0);
  return Math.round(clamped * BRIGHTNESS_SLIDER_FACTOR);
}

/**
 * 将滑块值转换为亮度值
 * @param {number} sliderValue - 滑块值（-100 到 100）
 * @returns {number} 亮度值（-1.0 到 1.0）
 */
export function sliderToBrightness(sliderValue) {
  const clamped = clamp(parseInt(sliderValue, 10), -100, 100, 0);
  return clamped / BRIGHTNESS_SLIDER_FACTOR;
}

/**
 * 将对比度值转换为滑块值
 * @param {number} contrast - 对比度值（-100 到 100）
 * @returns {number} 滑块值（-100 到 100）
 */
export function contrastToSlider(contrast) {
  return Math.round(clamp(contrast, -100, 100, 0));
}

/**
 * 将滑块值转换为对比度值
 * @param {number} sliderValue - 滑块值（-100 到 100）
 * @returns {number} 对比度值（-100 到 100）
 */
export function sliderToContrast(sliderValue) {
  return clamp(parseInt(sliderValue, 10), -100, 100, 0);
}

/**
 * 将饱和度值转换为滑块值
 * @param {number} saturation - 饱和度值（-100 到 100）
 * @returns {number} 滑块值（-100 到 100）
 */
export function saturationToSlider(saturation) {
  return Math.round(clamp(saturation, -100, 100, 0));
}

/**
 * 将滑块值转换为饱和度值
 * @param {number} sliderValue - 滑块值（-100 到 100）
 * @returns {number} 饱和度值（-100 到 100）
 */
export function sliderToSaturation(sliderValue) {
  return clamp(parseInt(sliderValue, 10), -100, 100, 0);
}

/**
 * 将透明度值转换为滑块值
 * @param {number} opacity - 透明度值（0 到 100）
 * @returns {number} 滑块值（0 到 100）
 */
export function opacityToSlider(opacity) {
  return Math.round(clamp(opacity, 0, 100, 100));
}

/**
 * 将滑块值转换为透明度值
 * @param {number} sliderValue - 滑块值（0 到 100）
 * @returns {number} 透明度值（0 到 100）
 */
export function sliderToOpacity(sliderValue) {
  return clamp(parseInt(sliderValue, 10), 0, 100, 100);
}

/**
 * 将透明度百分比转换为alpha值（0-1范围）
 * @param {number} opacity - 透明度百分比（0-100）
 * @returns {number} alpha值（0.0-1.0）
 */
export function opacityToAlpha(opacity) {
  const clamped = clamp(opacity, 0, 100, 100);
  return clamped / 100.0;
}

/**
 * 将alpha值转换为透明度百分比
 * @param {number} alpha - alpha值（0.0-1.0）
 * @returns {number} 透明度百分比（0-100）
 */
export function alphaToOpacity(alpha) {
  const clamped = clamp(alpha, 0.0, 1.0, 1.0);
  return clamped * 100.0;
}

/**
 * 获取默认的调节状态
 * @returns {Object} 默认调节状态
 */
export function getDefaultAdjustmentState() {
  return {
    brightness: ADJUSTMENT_RANGES.brightness.default,
    contrast: ADJUSTMENT_RANGES.contrast.default,
    saturation: ADJUSTMENT_RANGES.saturation.default,
    opacity: ADJUSTMENT_RANGES.opacity.default
  };
}

/**
 * 合并两个调节状态
 * @param {Object} baseState - 基础状态
 * @param {Object} overrideState - 覆盖状态
 * @returns {Object} 合并后的状态
 */
export function mergeAdjustmentStates(baseState = {}, overrideState = {}) {
  const normalizedBase = normalizeAdjustmentState(baseState);
  const normalizedOverride = normalizeAdjustmentState(overrideState);

  return {
    ...normalizedBase,
    ...normalizedOverride
  };
}

/**
 * 检查是否有激活的调节效果
 * @param {Object} state - 调节状态
 * @returns {boolean} 是否有激活的调节效果
 */
export function isAdjustmentActive(state = {}) {
  const normalized = normalizeAdjustmentState(state);
  const defaults = getDefaultAdjustmentState();

  return (
    Math.abs(normalized.brightness - defaults.brightness) > 0.001 ||
    Math.abs(normalized.contrast - defaults.contrast) > 0.001 ||
    Math.abs(normalized.saturation - defaults.saturation) > 0.001 ||
    Math.abs(normalized.opacity - defaults.opacity) > 0.001
  );
}

/**
 * 创建调节控件的配置
 * @returns {Object} 控件配置对象
 */
export function createAdjustmentControlsConfig() {
  return {
    brightness: {
      label: '亮度',
      min: -100,
      max: 100,
      defaultValue: 0,
      toSlider: brightnessToSlider,
      fromSlider: sliderToBrightness
    },
    contrast: {
      label: '对比度',
      min: -100,
      max: 100,
      defaultValue: 0,
      toSlider: contrastToSlider,
      fromSlider: sliderToContrast
    },
    saturation: {
      label: '饱和度',
      min: -100,
      max: 100,
      defaultValue: 0,
      toSlider: saturationToSlider,
      fromSlider: sliderToSaturation
    },
    opacity: {
      label: '透明度',
      min: 0,
      max: 100,
      defaultValue: 100,
      toSlider: opacityToSlider,
      fromSlider: sliderToOpacity
    }
  };
}

// 导出所有工具函数
export default {
  ADJUSTMENT_RANGES,
  BRIGHTNESS_SLIDER_FACTOR,
  clamp,
  normalizeAdjustmentState,
  brightnessToSlider,
  sliderToBrightness,
  contrastToSlider,
  sliderToContrast,
  saturationToSlider,
  sliderToSaturation,
  opacityToSlider,
  sliderToOpacity,
  opacityToAlpha,
  alphaToOpacity,
  getDefaultAdjustmentState,
  mergeAdjustmentStates,
  isAdjustmentActive,
  createAdjustmentControlsConfig
};

File: src/xiser_nodes/adjustment_utils.py
"""
adjustment_utils.py

统一的图像调节工具模块，提供参数范围定义、验证和转换函数。
确保前后端使用一致的参数范围和算法。
"""

import numpy as np
from PIL import Image


class AdjustmentUtils:
    """图像调节工具类"""

    # 统一的参数范围定义（与前端保持一致）
    BRIGHTNESS_RANGE = (-1.0, 1.0)          # 亮度范围：-1.0 到 1.0
    CONTRAST_RANGE = (-100.0, 100.0)        # 对比度范围：-100 到 100
    SATURATION_RANGE = (-100.0, 100.0)      # 饱和度范围：-100 到 100
    OPACITY_RANGE = (0.0, 100.0)            # 透明度范围：0 到 100

    # 默认值
    DEFAULT_BRIGHTNESS = 0.0
    DEFAULT_CONTRAST = 0.0
    DEFAULT_SATURATION = 0.0
    DEFAULT_OPACITY = 100.0

    @staticmethod
    def clamp(value, min_val, max_val, fallback=None):
        """
        限制值在指定范围内

        Args:
            value: 输入值
            min_val: 最小值
            max_val: 最大值
            fallback: 无效值时的默认值，如果为None则使用min_val和max_val的中间值

        Returns:
            float: 限制后的值
        """
        try:
            num = float(value)
            if np.isnan(num) or np.isinf(num):
                raise ValueError("Invalid number")
            return max(min_val, min(max_val, num))
        except (ValueError, TypeError):
            if fallback is not None:
                return fallback
            # 使用范围中间值作为默认值
            return (min_val + max_val) / 2.0

    @staticmethod
    def normalize_adjustment_state(state):
        """
        规范化调节状态，确保所有调节参数在有效范围内

        Args:
            state (dict): 原始调节状态

        Returns:
            dict: 规范化后的调节状态
        """
        if not isinstance(state, dict):
            state = {}

        return {
            "brightness": AdjustmentUtils.clamp(
                state.get("brightness", AdjustmentUtils.DEFAULT_BRIGHTNESS),
                *AdjustmentUtils.BRIGHTNESS_RANGE,
                AdjustmentUtils.DEFAULT_BRIGHTNESS
            ),
            "contrast": AdjustmentUtils.clamp(
                state.get("contrast", AdjustmentUtils.DEFAULT_CONTRAST),
                *AdjustmentUtils.CONTRAST_RANGE,
                AdjustmentUtils.DEFAULT_CONTRAST
            ),
            "saturation": AdjustmentUtils.clamp(
                state.get("saturation", AdjustmentUtils.DEFAULT_SATURATION),
                *AdjustmentUtils.SATURATION_RANGE,
                AdjustmentUtils.DEFAULT_SATURATION
            ),
            "opacity": AdjustmentUtils.clamp(
                state.get("opacity", AdjustmentUtils.DEFAULT_OPACITY),
                *AdjustmentUtils.OPACITY_RANGE,
                AdjustmentUtils.DEFAULT_OPACITY
            )
        }

    @staticmethod
    def opacity_to_alpha(opacity):
        """
        将透明度百分比转换为alpha值（0-1范围）

        Args:
            opacity (float): 透明度百分比（0-100）

        Returns:
            float: alpha值（0.0-1.0）
        """
        clamped = AdjustmentUtils.clamp(opacity, 0.0, 100.0, 100.0)
        return clamped / 100.0

    @staticmethod
    def alpha_to_opacity(alpha):
        """
        将alpha值转换为透明度百分比

        Args:
            alpha (float): alpha值（0.0-1.0）

        Returns:
            float: 透明度百分比（0-100）
        """
        clamped = AdjustmentUtils.clamp(alpha, 0.0, 1.0, 1.0)
        return clamped * 100.0

    @staticmethod
    def get_default_state():
        """
        获取默认的调节状态

        Returns:
            dict: 默认调节状态
        """
        return {
            "brightness": AdjustmentUtils.DEFAULT_BRIGHTNESS,
            "contrast": AdjustmentUtils.DEFAULT_CONTRAST,
            "saturation": AdjustmentUtils.DEFAULT_SATURATION,
            "opacity": AdjustmentUtils.DEFAULT_OPACITY
        }

    @staticmethod
    def merge_states(base_state, override_state):
        """
        合并两个调节状态，override_state中的值会覆盖base_state

        Args:
            base_state (dict): 基础状态
            override_state (dict): 覆盖状态

        Returns:
            dict: 合并后的状态
        """
        if not isinstance(base_state, dict):
            base_state = {}
        if not isinstance(override_state, dict):
            override_state = {}

        merged = base_state.copy()
        merged.update(override_state)
        return AdjustmentUtils.normalize_adjustment_state(merged)

    @staticmethod
    def is_adjustment_active(state):
        """
        检查是否有激活的调节效果

        Args:
            state (dict): 调节状态

        Returns:
            bool: 是否有激活的调节效果
        """
        if not isinstance(state, dict):
            return False

        normalized = AdjustmentUtils.normalize_adjustment_state(state)

        # 检查是否有非默认值
        return (
            abs(normalized["brightness"] - AdjustmentUtils.DEFAULT_BRIGHTNESS) > 1e-3 or
            abs(normalized["contrast"] - AdjustmentUtils.DEFAULT_CONTRAST) > 1e-3 or
            abs(normalized["saturation"] - AdjustmentUtils.DEFAULT_SATURATION) > 1e-3 or
            abs(normalized["opacity"] - AdjustmentUtils.DEFAULT_OPACITY) > 1e-3
        )


def create_adjustment_slider_config():
    """
    创建ComfyUI滑块控件的配置

    Returns:
        dict: 滑块控件配置
    """
    return {
        "brightness": ("FLOAT", {
            "default": AdjustmentUtils.DEFAULT_BRIGHTNESS,
            "min": AdjustmentUtils.BRIGHTNESS_RANGE[0],
            "max": AdjustmentUtils.BRIGHTNESS_RANGE[1],
            "step": 0.01,
            "display": "slider"
        }),
        "contrast": ("FLOAT", {
            "default": AdjustmentUtils.DEFAULT_CONTRAST,
            "min": AdjustmentUtils.CONTRAST_RANGE[0],
            "max": AdjustmentUtils.CONTRAST_RANGE[1],
            "step": 0.1,
            "display": "slider"
        }),
        "saturation": ("FLOAT", {
            "default": AdjustmentUtils.DEFAULT_SATURATION,
            "min": AdjustmentUtils.SATURATION_RANGE[0],
            "max": AdjustmentUtils.SATURATION_RANGE[1],
            "step": 0.1,
            "display": "slider"
        }),
        "opacity": ("FLOAT", {
            "default": AdjustmentUtils.DEFAULT_OPACITY,
            "min": AdjustmentUtils.OPACITY_RANGE[0],
            "max": AdjustmentUtils.OPACITY_RANGE[1],
            "step": 0.1,
            "display": "slider"
        })
    }